ui/src/table.tsx

```diff
@@ -727,35 +728,50 @@ export const
         const expandedRef = expandedRefs[key]
         return expandedRef === undefined || expandedRef
       },
+      initialGroups = React.useMemo(() => {
+        return m.groups
+          ? m.groups.reduce((acc, { label }) => {
+            acc[label] = ({ key: label, name: label, startIndex: 0, count: 0 })
+            return acc
+          }, {} as Dict<Fluent.IGroup>)
+          : groupByKey !== '*'
+            ? items.reduce((acc, item) => {
+              const group = (item as Fluent.IObjectWithKey & Dict<any>)[groupByKey]
+              if (!acc[group]) acc[group] = ({ key: group, name: group, startIndex: 0, count: 0 })
+              return acc
+            }, {} as Dict<Fluent.IGroup>)
+            : {}
+      }, [m.groups, groupByKey, items]),
```

Comment:
Wouldn't keeping group names only be enough?

```suggestion
      groupNames = React.useMemo(() => {
        return m.groups
          ? m.groups.reduce((acc, { label }) => acc.add(label), new Set<S>())
          : groupByKey !== '*'
            ? (items as Dict<S>[]).reduce((acc, item) => acc.add(item[groupByKey]), new Set<S>())
            : new Set<S>()
      }, [m.groups, groupByKey, items]),
```

Reply:
It is also an alternative.

Reply:
But you understand why is this better, right?

Reply:
It takes advantage of set - no need for checking if item exists before inserting.
It takes less memory to store names only.
It is shorter and easier to read.

Anything else?

Reply:
All correct. The last thing IIRC is that the previous solution mutated `useMemo` objects later on, which could lead into problems later on.